tests/integration/construct_component_failures/testcomponent/index.ts

```diff
@@ -0,0 +1,42 @@
+// Copyright 2016-2024, Pulumi Corporation.  All rights reserved.
+
+import * as pulumi from "@pulumi/pulumi";
+import * as provider from "@pulumi/pulumi/provider";
+import * as grpc from "@grpc/grpc-js";
+
+class Component extends pulumi.ComponentResource {
+    public readonly foo: pulumi.Output<string>;
+
+    constructor(name: string, foo: pulumi.Input<string>, opts?: pulumi.ComponentResourceOptions) {
+        super("testcomponent:index:Component", name, undefined, opts);
+
+        this.foo = pulumi.output(foo);
+
+        this.registerOutputs({
+            foo: this.foo,
+        })
+    }
+}
+
+class Provider implements provider.Provider {
+    public readonly version = "0.0.1";
+
+    async construct(name: string, type: string, inputs: pulumi.Inputs,
+              options: pulumi.ComponentResourceOptions): Promise<provider.ConstructResult> {
+        if (type != "testcomponent:index:Component") {
+            throw new Error(`unknown resource type ${type}`);
+        }
+
+        const component = new Component(name, inputs["foo"], options);
+
+	throw new pulumi.InvalidInputPropertiesError(
+	    "failing for a reason",
+	    [new pulumi.PropertyError("foo", "the failure reason")]);
```

Comment:
Sorry for not giving feedback sooner, but the usage of this seems kind of strange to me.

First, (and this is more of a nitpick for this test) for an actual component author, I'd expect them to throw the error from inside the component's constructor where the actual implementation lives. Construct shouldn't really have much implementation -- it's primarily just dispatching from the type token to the constructor.

Second, and more importantly, what does it look like for component authors to throw this error when validating multiple inputs? Right now, it seems like we're recommending that you'd keep track of multiple validations and roll them up into a single error? Is that typical of what we'd expect users to do?

I kinda would expect a pattern like this:

```typescript
constructor(name: string, args: ComponentArgs, opts?: pulumi.ComponentResourceOptions) {
    if (<args.foo is missing>) {
        throw new pulumi.InvalidInputPropertyError("foo", "foo is required");
    }
    if (<args.count isn't in range>) {
        throw new pulumi.InvalidInputPropertyError("count", "out of range");
    }
   ...
}
```

vs. what it looks like this design leans towards:

```typescript
constructor(name: string, args: ComponentArgs, opts?: pulumi.ComponentResourceOptions) {
    const errors: pulumi.PropertyError[] = [];

    if (<args.foo is missing>) {
        errors.push(new pulumi.PropertyError("foo", "foo is required"));
    }
    if (<args.count isn't in range>) {
        errors.push(new pulumi.PropertyError("count", "out of range"));
    }
    ...
    if (errors.length > 0) {
        throw new pulumi.InvalidInputPropertiesError("not sure what I should put here for an overall error message", errors);
    }
}
```

I get that the latter approach allows returning multiple errors to the user in one response, which means they don't have to fix-up issues one at a time piecemeal. But the usage (or at least the usage that the design is leading the user towards doing) from an author perspective seems unusual to me. I suppose they could still throw `InvalidInputPropertiesError` for each single case like in the former approach, but that feels like I'm using the API incorrectly based on the naming and signature of `InvalidInputPropertiesError`.

Reply:
Could we discuss the details of this more in ideation before merging?

Reply:
One additional aspect of this that I'm curious about, but haven't dug into the details of... For components, the input properties can be instances of `Output<T>` and `apply` needs to be used to access the actual values to do any validation. How would that work with this? Would throwing inside the `apply` propagate through as expected?

Reply:
Happy to discuss this in ideation.  The idea of having multiple errors being returned here comes from @EronWright's original PR (https://github.com/pulumi/pulumi/pull/16132), so I'd be curious about his take here as well.

Reply:
I wonder if the best approach here is to just allow both, and have a separate `InvalidInputPropertyError` that can be thrown, while also having the `InvalidInputPropertiesError`, in case the provider wants to indicate multiple problems at once.  The latter is clearly more user friendly if there are multiple errors, and I think the end user experience is probably what we should optimize most ahead of the experience of a component provider (I'm assuming the former is more frequent than the latter).

Reply:
I was simply following the example set by `Check` validation errors, not trying to be creative.